src/solver/reduce/forward_sum.cpp

```diff
@@ -0,0 +1,334 @@
+    {
+        MIOPEN_THROW(miopenStatusBadParm, "SumForward: Reduce last dim not supported.");
+    }
+    return true;
+}
+
+bool IsImprovementOverROCm(const ExecutionContext& context,
+                           const miopen::reduce::ProblemDescription& problem)
+{
+    auto xdims = problem.GetXDesc().GetLengths();
+    auto ydims = problem.GetYDesc().GetLengths();
+    auto dim   = problem.GetDim();
+
+    auto reduce_size = xdims[dim];
+    auto output_numel =
+        std::accumulate(ydims.begin(), ydims.end(), 1ULL, std::multiplies<size_t>());
+
+    // At least 4 WGs per one CU
+    auto reqd_work_item_cnt =
+        static_cast<size_t>(256 * context.GetStream().GetMaxComputeUnits() * 4);
+
+    bool is_num_work_item_enough = (output_numel > reqd_work_item_cnt);
+    bool is_parallelism_enough   = (output_numel * reduce_size > reqd_work_item_cnt);
+
+    if(!is_num_work_item_enough && is_parallelism_enough)
+    {
+        // It's large enough to parallelization, but calling the kernel twice is overhead.
+        // For cases smaller than this, ROCm pytorch performed better.
+        bool is_improvement_ROCm = (output_numel * reduce_size < reqd_work_item_cnt * 64);
+        // But the reduce size is small, MIOpen HIP performs better.
+        bool is_reduce_large = (reduce_size > 64);
+
+        if(is_improvement_ROCm && is_reduce_large)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "SumForward: Performance cannot be guaranteed.");
+        }
+    }
+
+    return true;
+}
+
+bool SumForward::IsApplicable(const ExecutionContext& context,
+                              const miopen::reduce::ProblemDescription& problem) const
+{
+    if(!problem.IsSameType())
+        return false;
+    if(!problem.IsRightDim())
+        return false;
+    if(!problem.IsRightLength())
+        return false;
+    if(!problem.IsAllPacked())
+        return false;
+    if(!IsNotLastDim(problem))
+        return false;
+    if(!IsImprovementOverROCm(context, problem))
+        return false;
+    return true;
+}
+
+ConvSolution SumForward::GetSolution(const ExecutionContext& context,
+                                     const miopen::reduce::ProblemDescription& problem) const
+{
+    std::ignore = context;
+
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto dtype = problem.GetXDesc().GetType();
+    auto xdims = problem.GetXDesc().GetLengths();
+    auto ydims = problem.GetYDesc().GetLengths();
+    auto dim   = problem.GetDim();
+
+    auto reduce_size = xdims[dim];
+    auto output_numel =
+        std::accumulate(ydims.begin(), ydims.end(), 1ULL, std::multiplies<size_t>());
+
+    // At least 4 WGs per one CU
+    auto reqd_work_item_cnt =
+        static_cast<size_t>(256 * context.GetStream().GetMaxComputeUnits() * 4);
+
+    bool is_num_work_item_enough = (output_numel > reqd_work_item_cnt);
+    bool is_parallelism_enough   = (output_numel * reduce_size > reqd_work_item_cnt);
```

Comment:
The same code is used inside `IsImprovementOverROCm` and inside `GetWorkspaceSize`.

Reply:
fixed it